pkg/detector/detector.go

```diff
@@ -373,7 +373,7 @@ func (d *ResourceDetector) LookForMatchedClusterPolicy(object *unstructured.Unst
 }
 
 // ApplyPolicy starts propagate the object referenced by object key according to PropagationPolicy.
-func (d *ResourceDetector) ApplyPolicy(object *unstructured.Unstructured, objectKey keys.ClusterWideKey, policy *policyv1alpha1.PropagationPolicy) (err error) {
+func (d *ResourceDetector) ApplyPolicy(object *unstructured.Unstructured, objectKey keys.ClusterWideKey, policy *policyv1alpha1.PropagationPolicy, removedLabelKeys ...string) (err error) {
```

Comment:
Just feel weird about adding `removedLabelKeys` to `ApplyPolicy`. `ApplyPolicy` should not take the responsibility of `preemption`.

A generally idea is that can we handle preemption by replace the `claimed info`?
For example, when a PP preempts a CPP:
- 1. Remove the claimed label(`clusterpropagationpolicy.karmada.io/name`) from the resource template
- 2. Claim the resource template by PP's label (`clusterpropagationpolicy.karmada.io/name`)

Reply:
> For example, when a PP preempts a CPP:

> *  i.Remove the claimed label(clusterpropagationpolicy.karmada.io/name) from the resource template
> *  ii. Claim the resource template by PP's label (clusterpropagationpolicy.karmada.io/name)

1. the claimed label `clusterpropagationpolicy.karmada.io/name` on rb cannot be removed if we follow the above steps.
2. if we remove the claimed label `clusterpropagationpolicy.karmada.io/name` from the resource template and rb, and then claim new pp label on them. there will be two update operations, seems uncessary.

Reply:
BTW, the responsibility of `ApplyPolicy` is to apply the new policy, and I think it also includes the semantics of removing the old policy.

Reply:
> the claimed label clusterpropagationpolicy.karmada.io/name on rb cannot be removed if we follow the above steps.

That's a thing we have to handle.

>  there will be two update operations, seems necessary.

We can do that by updating once.

Reply:
so you just don't think it's appropriate to remove the old policy label and apply the new policy through `ApplyPolicy`, right?
We should create a new function to solve this problem, giving it a new meaning.

Reply:
Yes,  It'd be better not to couple the original logic. 

We can make a simple preemption function that focuses on preempting by changing the claim info on the resource template.
After that, the resource template would get a chance to reconcile, that is the previous apply process.

Reply:
> After that, the resource template would get a chance to reconcile, that is the previous apply process.

I'm afraid it won't work as expected if just update the resource template and have it reconcile again.
https://github.com/karmada-io/karmada/blob/cf90ffc64fbfa16083c8371905b4789fb4f98a6f/pkg/detector/detector.go#L394-L414

finally, binding will be like:
```
apiVersion: work.karmada.io/v1alpha2
kind: ResourceBinding
metadata:
  creationTimestamp: "2023-07-28T06:36:01Z"
  finalizers:
  - karmada.io/binding-controller
  generation: 2
  labels:
    clusterpropagationpolicy.karmada.io/name: xxxx
    propagationpolicy.karmada.io/name: yyyy
    propagationpolicy.karmada.io/namespace: default
```
so, the possible way is to provide a function similar to `ApplyPolicy` but only for preemption. how do you think of this?

Reply:
> the possible way is to provide a function similar to ApplyPolicy but only for preemption. 

We'd better not. 

When building(updating) a RB, we know which claim label should exist at :
https://github.com/karmada-io/karmada/blob/cf90ffc64fbfa16083c8371905b4789fb4f98a6f/pkg/detector/detector.go#L394-L414

Reply:
function `util.DedupeAndMergeLabels` only merge new labels into old binding, not remove label `clusterpropagationpolicy.karmada.io/name`, which is the root of the problem. so is it necessary for us to change the behavior here?
```
bindingCopy.Labels = util.DedupeAndMergeLabels(bindingCopy.Labels, binding.Labels) 
```

Reply:
Here is an ugly solution:
```diff
diff --git a/pkg/detector/detector.go b/pkg/detector/detector.go                                     
index be715e631..86695f277 100644                                                                    
--- a/pkg/detector/detector.go                                                                       
+++ b/pkg/detector/detector.go                                                                       
@@ -422,6 +422,11 @@ func (d *ResourceDetector) ApplyPolicy(object *unstructured.Unstructured, object
                        bindingCopy.Spec.Placement = binding.Spec.Placement                          
                        bindingCopy.Spec.Failover = binding.Spec.Failover                            
                        bindingCopy.Spec.ConflictResolution = binding.Spec.ConflictResolution        
+                                                                                                    
+                       // cleanup in case of current ResourceBinding is preempted from
+                       // a ClusterPropagationPolicy
+                       delete(binding.Labels, policyv1alpha1.ClusterPropagationPolicyLabel)
+
                        return nil
                })
                if err != nil {

```